File: gtci/02-fast-and-slow-pointers/python3/6-rearrange-a-linked-list.py
class Node:
    def __init__(self, value, next=None):
        self.val = value
        self.next = next


class Solution:
    def reorder(self, head):
        middle = self.get_middle_node(head)

        first_half_curr = head
        second_half_curr = self.reverse_list(middle)

        while first_half_curr is not None and second_half_curr is not None:
            temp = first_half_curr.next

            first_half_curr.next = second_half_curr
            first_half_curr = temp

            temp = second_half_curr.next

            second_half_curr.next = first_half_curr
            second_half_curr = temp

        if first_half_curr is not None:
            first_half_curr.next = None

        return head

    def get_middle_node(self, head):
        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next

        return slow

    def reverse_list(self, head):
        prev = None
        curr = head

        while curr is not None:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        return prev

    def get_list_values(self, head):
        values = []
        curr = head
        while curr:
            values.append(curr.val)
            curr = curr.next
        return values


if __name__ == "__main__":
    # Test case 1: Even length list [1, 2, 3, 4]
    # Expected result: [1, 4, 2, 3]
    head1 = Node(1)
    head1.next = Node(2)
    head1.next.next = Node(3)
    head1.next.next.next = Node(4)

    sol = Solution()
    reordered1 = sol.reorder(head1)
    assert sol.get_list_values(reordered1) == [1, 4, 2, 3], "Test case 1 failed"

    # Test case 2: Odd length list [1, 2, 3, 4, 5]
    # Expected result: [1, 5, 2, 4, 3]
    head2 = Node(1)
    head2.next = Node(2)
    head2.next.next = Node(3)
    head2.next.next.next = Node(4)
    head2.next.next.next.next = Node(5)

    reordered2 = sol.reorder(head2)
    assert sol.get_list_values(reordered2) == [1, 5, 2, 4, 3], "Test case 2 failed"

    # Test case 3: Short list [1, 2]
    # Expected result: [1, 2]
    head3 = Node(1)
    head3.next = Node(2)

    reordered3 = sol.reorder(head3)
    assert sol.get_list_values(reordered3) == [1, 2], "Test case 3 failed"

    # Test case 4: Single node [1]
    # Expected result: [1]
    head4 = Node(1)

    reordered4 = sol.reorder(head4)
    assert sol.get_list_values(reordered4) == [1], "Test case 4 failed"

    print("All test cases passed.")

File: gtci/02-fast-and-slow-pointers/cpp/6-rearrange-a-linked-list.cpp
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

class ListNode {
   public:
    int val = 0;
    ListNode* next;

    ListNode(int value) {
        this->val = value;
        next = nullptr;
    }
};

class Solution {
   public:
    ListNode* reorder(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }

        ListNode* middle = getMiddleNode(head);
        ListNode* firstHalfCurr = head;
        ListNode* secondHalfCurr = reverseList(middle);

        while (firstHalfCurr != nullptr && secondHalfCurr != nullptr) {
            ListNode* temp = firstHalfCurr->next;
            firstHalfCurr->next = secondHalfCurr;
            firstHalfCurr = temp;

            temp = secondHalfCurr->next;
            secondHalfCurr->next = firstHalfCurr;
            secondHalfCurr = temp;
        }

        if (firstHalfCurr != nullptr) {
            firstHalfCurr->next = nullptr;
        }

        return head;
    }

    vector<int> getListValues(ListNode* head) {
        vector<int> values;
        ListNode* curr = head;
        while (curr != nullptr) {
            values.push_back(curr->val);
            curr = curr->next;
        }
        return values;
    }

   private:
    ListNode* getMiddleNode(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }

    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;
        while (curr != nullptr) {
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }
        return prev;
    }
};

int main() {
    // Test case 1: Even length list [1, 2, 3, 4]
    // Expected result: [1, 4, 2, 3]
    ListNode* head1 = new ListNode(1);
    head1->next = new ListNode(2);
    head1->next->next = new ListNode(3);
    head1->next->next->next = new ListNode(4);

    Solution sol;
    ListNode* reordered1 = sol.reorder(head1);
    vector<int> expected1 = {1, 4, 2, 3};
    assert(sol.getListValues(reordered1) == expected1);

    // Test case 2: Odd length list [1, 2, 3, 4, 5]
    // Expected result: [1, 5, 2, 4, 3]
    ListNode* head2 = new ListNode(1);
    head2->next = new ListNode(2);
    head2->next->next = new ListNode(3);
    head2->next->next->next = new ListNode(4);
    head2->next->next->next->next = new ListNode(5);

    ListNode* reordered2 = sol.reorder(head2);
    vector<int> expected2 = {1, 5, 2, 4, 3};
    assert(sol.getListValues(reordered2) == expected2);

    // Test case 3: Short list [1, 2]
    // Expected result: [1, 2]
    ListNode* head3 = new ListNode(1);
    head3->next = new ListNode(2);

    ListNode* reordered3 = sol.reorder(head3);
    vector<int> expected3 = {1, 2};
    assert(sol.getListValues(reordered3) == expected3);

    // Test case 4: Single node [1]
    // Expected result: [1]
    ListNode* head4 = new ListNode(1);

    ListNode* reordered4 = sol.reorder(head4);
    vector<int> expected4 = {1};
    assert(sol.getListValues(reordered4) == expected4);

    cout << "All test cases passed." << endl;

    // Clean up memory
    for (ListNode* curr = head1; curr != nullptr;) {
        ListNode* temp = curr;
        curr = curr->next;
        delete temp;
    }
    for (ListNode* curr = head2; curr != nullptr;) {
        ListNode* temp = curr;
        curr = curr->next;
        delete temp;
    }
    for (ListNode* curr = head3; curr != nullptr;) {
        ListNode* temp = curr;
        curr = curr->next;
        delete temp;
    }
    for (ListNode* curr = head4; curr != nullptr;) {
        ListNode* temp = curr;
        curr = curr->next;
        delete temp;
    }

    return 0;
}
